### PR TITLE
[FIRRTL][IST] Verify inner symbols are unique.

### LINKED ISSUE #3531: [FIRRTL] Report error on duplicate inner symbols (IST verifier?)
## Problem

In FIRRTL, inner symbols can be repeated without the verifier objecting.

### Example

Here's an example:

**syms.mlir**:
```mlir
module {
  firrtl.circuit "Syms" {
    firrtl.module @Syms() {
      %w = firrtl.wire sym @x : !firrtl.uint<2>
      %w2 = firrtl.wire sym @x : !firrtl.uint<2>
      %c0_i2 = firrtl.constant 0 : !firrtl.uint<2>
      firrtl.strictconnect %w, %c0_i2 : !firrtl.uint<2>
      firrtl.strictconnect %w2, %w : !firrtl.uint<2>
    }
    sv.verbatim "Testing InnerSym ref to '@Syms::@x': {{0}}" { symbols = [ #hw.innerNameRef<@Syms::@x> ] }
  }
}
```

Which with assertions disabled, firtool happily processes and produces:

```
module Syms();  // ./syms.mlir:3:5
  wire [1:0] w = 2'h0;  // ./syms.mlir:4:12, :6:16
  wire [1:0] w2;        // ./syms.mlir:5:13

  assign w2 = w;        // ./syms.mlir:8:7
endmodule

Testing InnerSym ref to '@Syms::@x': w  // ./syms.mlir:10:5
```

There *is* an assert that catches this, but it's while constructing IST's for purposes of verifying InnerRef's.  This would be good to be verifying explicitly and should produce a proper error message.

The assert at least means we're not likely to be encountering this in practice, I believe.

## Proposed Solution(s)

One solution is to have a verifier for InnerSymbolTable (the trait, "IST").  This would check that all symbols are unique.

Verifying this on the IST trait seems conceptually appropriate, but that means yet another place we're walking modules to check symbols which we're trying to do less frequently not more.  Note that we can't share the constructed IST from Passes or even between verifiers (AFAIK).  Pushing this up into the IRN (InnerRefNamespace) verifier (circuit-level) as it builds the IST's would be more efficient.

## Related
cc #3526, which is about similar issue in HW.

Presently we have a verifier for checking that all inner symbols within an operation (fields) are unique.

### PR DESCRIPTION
**Pull request by dtzSiFive**
Split construction into verification paths.

Do verification as part of circuit-level inner symbol verification, which is building all the IST's already (and is required to be parent of any IST).

Add some tests for duplicate symbol errors.

Fixes #3531.

### CODE REVIEW

File: lib/Dialect/FIRRTL/InnerSymbolTable.cpp [hunk 1]
```diff
@@ -186,17 +223,19 @@ Operation *InnerRefNamespace::lookupOp(hw::InnerRefAttr inner) {
...
-LogicalResult verifyInnerRefs(Operation *op) {
+LogicalResult verifyInnerRefNamespace(Operation *op) {
   // Construct the symbol tables.
   InnerSymbolTableCollection innerSymTables;
   SymbolTable symbolTable(op);
   InnerRefNamespace ns{symbolTable, innerSymTables};
-  innerSymTables.populateTables(op);
+
+  if (failed(innerSymTables.populateAndVerifyTables(op)))
+    return failure();
```
Comment (youngar) [inline lib/Dialect/FIRRTL/InnerSymbolTable.cpp:238]: Makes more sense to me to populate the table before constructing the `InnerRefNamespace` from it.

File: lib/Dialect/FIRRTL/InnerSymbolTable.cpp [hunk 2]
```diff
@@ -186,17 +223,19 @@ Operation *InnerRefNamespace::lookupOp(hw::InnerRefAttr inner) {
...
+// InnerRefNamespace verification
 //===----------------------------------------------------------------------===//
 
 namespace detail {
 
-LogicalResult verifyInnerRefs(Operation *op) {
+LogicalResult verifyInnerRefNamespace(Operation *op) {
   // Construct the symbol tables.
   InnerSymbolTableCollection innerSymTables;
   SymbolTable symbolTable(op);
```
Comment (youngar) [inline lib/Dialect/FIRRTL/InnerSymbolTable.cpp:234]: Should the `SymbolTable` verifier run before the `InnerRefNamespace` verifier?  https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/OpBase.td#L1732

**Review by youngar — APPROVED** [review]
LGTM

File: lib/Dialect/FIRRTL/InnerSymbolTable.cpp [hunk 3]
```diff
@@ -26,54 +26,87 @@ namespace firrtl {
...
-    assert(it.second && "repeated symbol found");
+    return callback(name, target);
   };
-  auto addSyms = [&](InnerSymAttr symAttr, InnerSymTarget baseTarget) {
+
+  auto addSyms = [&](InnerSymAttr symAttr,
+                     InnerSymTarget baseTarget) -> LogicalResult {
     if (!symAttr)
-      return;
+      return success();
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/InnerSymbolTable.cpp:80]: Nope, but the InnerSymAttr itself might be null.  Seemed neater to check in the lambda than caller-side.  That said, it'd save constructing InnerSymTarget's just to drop them (regardless of whether that gets compiled away), so I'll give that a go :+1: .

File: lib/Dialect/FIRRTL/InnerSymbolTable.cpp [hunk 4]
```diff
@@ -26,54 +26,87 @@ namespace firrtl {
...
+
+LogicalResult InnerSymbolTable::walkSymbols(Operation *op,
+                                            InnerSymCallbackFn callback) {
+  using llvm::dbgs;
+  LLVM_DEBUG(dbgs() << "===----- InnerSymbolTable -----===\n";
+             dbgs() << "Walking inner symbols for @"
+                    << SymbolTable::getSymbolName(op) << "\n");
   auto addSym = [&](StringAttr name, InnerSymTarget target) {
     LLVM_DEBUG(dbgs() << " - @" << name << " -> " << target << "\n");
     assert(name && !name.getValue().empty());
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/InnerSymbolTable.cpp:73]: Assert liberally! But yes it would be better for them to be correct by construction.  This was moreso to catch my own mistake than accommodate invalid Attr's.  Although presently we have empty strings for ports that don't have symbols (if at least one port has a valid symbol, else the array is empty).

Maybe more reason to push this into InnerSymbolOpInterface / IST methods that callback with valid symbols/targets, to keep those details isolated there!

File: lib/Dialect/FIRRTL/InnerSymbolTable.cpp [hunk 5]
```diff
@@ -26,54 +26,87 @@ namespace firrtl {
...
+               return WalkResult::interrupt();
+
+           // Check for ports
+           // TODO: Add fields per port, once they work that way (use addSyms)
+           if (auto mod = dyn_cast<FModuleLike>(curOp)) {
+             for (const auto &p : llvm::enumerate(mod.getPorts()))
+               if (auto sym = p.value().sym; sym && !sym.getValue().empty())
+                 if (failed(addSym(p.value().sym,
+                                   InnerSymTarget(p.index(), curOp))))
+                   return WalkResult::interrupt();
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/InnerSymbolTable.cpp:106]: Hmm, not sure I follow, mostly because still need to know what port (index) the symbol is for (and because apparently that returns an `ArrayRef<Attribute>` not `ArrayRef<StringAttr>` (or `ArrayRef<InnerSymAttr>`)).  Here's the sort of thing I ended up w/`getPortSymbols`:

```
for (const auto &p : llvm::enumerate(mod.getPortSymbols()))
  if (auto sym = p.value().cast<StringAttr>();
      sym && !sym.getValue().empty() /* quirk of port symbols */)
if (failed(walkSym(sym, InnerSymTarget(p.index(), curOp))))
  return WalkResult::interrupt();
```

Maybe this means we need better way to get port symbols.. :+1: .

File: lib/Dialect/FIRRTL/InnerSymbolTable.cpp (hunk 1 above)
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/InnerSymbolTable.cpp:232]: Good call, thanks!

File: lib/Dialect/FIRRTL/InnerSymbolTable.cpp [hunk 6]
```diff
@@ -26,54 +26,87 @@ namespace firrtl {
...
-      for (const auto &p : llvm::enumerate(mod.getPorts()))
-        if (auto sym = p.value().sym; sym && !sym.getValue().empty())
-          addSym(p.value().sym, InnerSymTarget(p.index(), curOp));
-    }
-  });
+  return success(
+      !op->walk<mlir::WalkOrder::PreOrder>([&](Operation *curOp) -> WalkResult {
+           if (auto symOp = dyn_cast<InnerSymbolOpInterface>(curOp))
+             if (failed(
+                     addSyms(symOp.getInnerSymAttr(), InnerSymTarget(symOp))))
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/InnerSymbolTable.cpp]: Agreed, missed this thanks!

File: lib/Dialect/FIRRTL/InnerSymbolTable.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (youngar) [reply lib/Dialect/FIRRTL/InnerSymbolTable.cpp]: Yeah I was mostly thinking that `getPorts` seems kind of expensive, since it constructs a small vector with lots of different information about each port.  Did you try the similar `getPortSymbolsAttr().getAsRange<StringAttr>()`? (does that still work with enums?)

Another idea is to use a regular for loop:
```
for (int i = 0, e = numPorts; i < e) {
  if (failed(addSym(mod.getPortSymbolAttr(i);
                                   InnerSymTarget(i, curOp))))
                   return WalkResult::interrupt();
}
```

[the remark this answers is not in the document]
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/InnerSymbolTable.cpp]: Thanks!
